website/registrations/utils.py

```diff
@@ -0,0 +1,445 @@
+                               'Read-Only Contributors',
+                               'Bibliographic Contributors']
+
+CATEGORY_REVERSE_LOOKUP = {display_name: name for name, display_name in settings.NODE_CATEGORY_MAP.items()}
+
+MAX_EXCEL_COLUMN_NUMBER = 16384
+
+@functools.lru_cache(maxsize=MAX_EXCEL_COLUMN_NUMBER)
+def get_excel_column_name(column_index):
+    column_name = ''
+    # Fix zero indexing
+    current_column = column_index + 1
+
+    while current_column > 0:
+        modulo = (current_column - 1) % 26
+        column_name = '{}{}'.format(string.ascii_uppercase[modulo], column_name)
+        current_column = floor((current_column - modulo) / 26)
+
+    return column_name
+
+class Store():
+    def __init__(self, registration_provider):
+        self.licenses = registration_provider.licenses_acceptable.all()
+        self.subjects = registration_provider.all_subjects.all()
+        self.institutions = Institution.objects.get_all_institutions()
+
+class InvalidHeadersError(ValidationError):
+    pass
+
+class BulkRegistrationUpload():
+    @property
+    def is_valid(self):
+        return self.is_validated and not len(self.errors)
+
+    @property
+    def is_validated(self):
+        return all([row.is_validated for row in self.rows])
+
+    def __init__(self, bulk_upload_csv, provider_id):
+        self.raw_csv = bulk_upload_csv.read().decode('utf-8')
+        csv_io = io.StringIO(self.raw_csv)
+        self.reader = csv.DictReader(csv_io)
+        self.headers = self.reader.fieldnames
+        schema_id_row = next(self.reader)
+        self.schema_id = schema_id_row[self.reader.fieldnames[0]]
+        self.provider_id = provider_id
+
+        self.registration_provider = RegistrationProvider.load(self.provider_id)
+        if self.registration_provider is None:
+            raise NotFound(detail='Registration provider with id "{}" was not found').format(self.provider_id)
+
+        try:
+            self.registration_schema = self.registration_provider.schemas.get(_id=self.schema_id)
+        except RegistrationSchema.DoesNotExist:
+            raise NotFound(detail='Schema with id "{}" was not found'.format(self.schema_id))
+        self.schema_questions = BulkRegistrationUpload.get_schema_questions_validations(self.registration_schema)
+        self.validations = {**self.schema_questions, **METADATA_FIELDS}
+        self.errors = []
+        self.validate_csv_header_list()
+        self.store = Store(self.registration_provider)
+        self.rows = [Row(row,
+                         self.validations,
+                         self.store,
+                         functools.partial(self.log_error, row_index=index + 3))
+                     for index, row in enumerate(self.reader)]
+        csv_io.close()
+
+    @classmethod
+    def get_schema_questions_validations(cls, registration_schema):
+        get_multiple_choice_options = lambda schema_block, schema_blocks: [
```

Comment:
Did some homework because Matt asked, and want to add that CL Long Form schema has 112 `*-input` or `select-input-option` blocks while CL Short Form has 100, so the change from O(n^2) to O(n) is non-trivial for our main use case.